crates/executor/src/executor.rs

```diff
@@ -17,258 +19,423 @@ use revm::{
+                //e => return Err(Error::EVMError { error_code: e as u32 }),
+            };
+
+            // Add spent gas.
+            cumulative_gas_used += gas_used;
+
+            // commit state
+            let (changeset, new_bytecodes) = self.commit_changes(state);
+
+            // Transform logs to reth format.
+            let logs: Vec<Log> = logs.into_iter().map(into_reth_log).collect();
+
+            // Push transaction changeset and calculte header bloom filter for receipt.
+            changesets.push(TransactionChangeSet {
+                receipt: Receipt {
+                    tx_type: transaction.tx_type(),
+                    success: is_success,
+                    cumulative_gas_used,
+                    bloom: logs_bloom(logs.iter()),
+                    logs,
+                },
+                changeset,
+                new_bytecodes,
+            })
+        }
+
+        // Check if gas used matches the value set in header.
+        if header.gas_used != cumulative_gas_used {
+            return Err(Error::BlockGasUsed { got: cumulative_gas_used, expected: header.gas_used })
+        }
+
+        let mut block_reward = self.block_reward_changeset(header, ommers)?;
+
+        if self.chain_spec.fork_block(Hardfork::Dao) == Some(header.number) {
+            let mut irregular_state_changeset = self.dao_fork_changeset()?;
+            irregular_state_changeset.extend(block_reward.take().unwrap_or_default().into_iter());
+            block_reward = Some(irregular_state_changeset);
+        }
+
+        Ok(ExecutionResult { changesets, block_reward })
+    }
 }
 
 /// Execute and verify block
 pub fn execute_and_verify_receipt<DB: StateProvider>(
-    header: &Header,
-    transactions: &[TransactionSignedEcRecovered],
-    ommers: &[Header],
+    block: &Block,
+    senders: Option<Vec<Address>>,
     chain_spec: &ChainSpec,
     db: &mut SubState<DB>,
 ) -> Result<ExecutionResult, Error> {
-    let transaction_change_set = execute(header, transactions, ommers, chain_spec, db)?;
+    let execution_result = execute(block, senders, chain_spec, db)?;
 
-    let receipts_iter =
-        transaction_change_set.changesets.iter().map(|changeset| &changeset.receipt);
+    let receipts_iter = execution_result.changesets.iter().map(|changeset| &changeset.receipt);
 
-    if Some(header.number) >= chain_spec.fork_block(Hardfork::Byzantium) {
-        verify_receipt(header.receipts_root, header.logs_bloom, receipts_iter)?;
+    if Some(block.header.number) >= chain_spec.fork_block(Hardfork::Byzantium) {
+        verify_receipt(block.header.receipts_root, block.header.logs_bloom, receipts_iter)?;
     }
     // TODO Before Byzantium, receipts contained state root that would mean that expensive operation
     // as hashing that is needed for state root got calculated in every transaction
     // This was replaced with is_success flag.
     // See more about EIP here: https://eips.ethereum.org/EIPS/eip-658
```

Comment:
@rakita how does geth do it when syncing from genesis? what about erigon?

Reply:
moving discussion here: https://github.com/paradigmxyz/reth/issues/853